xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -949,18 +929,691 @@ func (s) TestRingHash_EndpointWeights(t *testing.T) {
+	backends := startTestServiceBackends(t, 2)
+	const clusterName = "cluster"
+	endpoints := endpointResource(t, clusterName, backends)
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		ServiceName: clusterName,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+
+	route := e2e.DefaultRouteConfig("new_route", "test.server", clusterName)
+
+	// Even though this policy is terminal, since it produces no result, we
+	// continue past it to find a policy that produces results.
+	hashPolicy := v3routepb.RouteAction_HashPolicy{
+		PolicySpecifier: &v3routepb.RouteAction_HashPolicy_Header_{
+			Header: &v3routepb.RouteAction_HashPolicy_Header{
+				HeaderName: "fixed_string",
+			},
+		},
+		Terminal: true,
+	}
+	hashPolicy2 := v3routepb.RouteAction_HashPolicy{
+		PolicySpecifier: &v3routepb.RouteAction_HashPolicy_Header_{
+			Header: &v3routepb.RouteAction_HashPolicy_Header{
+				HeaderName: "random_string",
+			},
+		},
+	}
+	action := route.VirtualHosts[0].Routes[0].Action.(*v3routepb.Route_Route)
+	action.Route.HashPolicy = []*v3routepb.RouteAction_HashPolicy{&hashPolicy, &hashPolicy2}
+
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	xdsServer, nodeID, xdsResolver := setupManagementServerAndResolver(t)
+	if err := xdsServer.Update(ctx, xdsUpdateOpts(nodeID, endpoints, cluster, route, listener)); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	// Check that despite the matching random string header, since the fixed
+	// string hash policy is terminal, requests all get routed to the same
+	// host.
+	gotPerBackend := make(map[string]int)
+	const numRPCs = 100
+	for i := 0; i < numRPCs; i++ {
+		ctx := metadata.NewOutgoingContext(ctx, metadata.Pairs(
+			"fixed_string", backends[0]+"_0",
+			"random_string", fmt.Sprintf("%d", rand.Int())),
+		)
+		var remote peer.Peer
+		_, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&remote))
+		if err != nil {
+			t.Fatalf("rpc EmptyCall() failed: %v", err)
+		}
+		gotPerBackend[remote.Addr.String()]++
+	}
+
+	if len(gotPerBackend) != 1 {
+		t.Error("Got RPCs routed to multiple backends, want a single backend")
+	}
+	if got := gotPerBackend[backends[0]]; got != numRPCs {
+		t.Errorf("Got %v RPCs routed to %v, want %v", got, backends[0], numRPCs)
+	}
+}
+
+// TestRingHash_IdleToReady tests that the channel will go from idle to ready
+// via connecting; (though it is not possible to catch the connecting state
+// before moving to ready).
+func (s) TestRingHash_IdleToReady(t *testing.T) {
```

Comment:
We have an internal API that allows one to subscribe to connectivity state changes on the channel. See: https://github.com/grpc/grpc-go/blob/f199062ef31ddda54152e1ca5e3d15fb63903dc3/internal/internal.go#L139

Do you think it makes sense to use that for this test?

Reply:
Interesting. That would allow us to see all transitions. But this is an internal API that is unused within the project. Do you have some usages within Google, since IIUC you don't enforce internal there? I would rather not depend on an API that is internal & unused, and would perhaps even propose to remove it.

Reply:
There are two places we intend to use this eventually within the codebase, but just haven't found time to do it.
- Monitoring the state of the xDS client channel. We currently get around this by using `WaitForReady` call option in the ADS stream call.
- Monitoring the state of the RLS control channel. We currently have some complicated logic to achieve this and it makes the associated test a little flaky. It's been on my wish list for a while to switch RLS to use this internal API and make that test non-flaky.

So, I'd be happy if we end up using it here and see some usage for this API. But I'm totally OK if you dont want to use it as well. Thanks.

Reply:
If you prefer not to add that in as part of this PR, maybe adding a TODO to use the internal API would work as well. Thanks.

Reply:
OK, I added a TODO. Using the SubscribeToConnectivityStateChanges API turned out to be a bit verbose and I think it's OK if we miss some state changes in this test.